timeline_sync/api.py

```diff
@@ -0,0 +1,172 @@
+        # TODO: return 404 if app does not have timeline support or user is not authorised in dev portal
+        sandbox_token = SandboxToken(user_id=uid, app_uuid=app_uuid, token=secrets.token_urlsafe(32))
+        db.session.add(sandbox_token)
+        db.session.commit()
+
+    result = {"uuid": sandbox_token.app_uuid, "token": sandbox_token.token}
+
+    return jsonify(result)
+
+
+@api.route('/sync')
+def sync():
+    user_id = get_uid()
+
+    user_timeline = db.session.query(UserTimeline).filter_by(user_id=user_id)
+
+    updates = [user_timeline_item.to_json() for user_timeline_item in user_timeline]
+
+    user_timeline.delete()
+    db.session.commit()
+
+    result = {
+        "updates": updates,
+        "syncURL": url_for('api.sync', _external=True)
+    }
+    return jsonify(result)
+
+
+@api.route('/user/pins/<pin_id>', methods=['PUT', 'DELETE'])
+def user_pin(pin_id):
+    try:
+        user_token = request.headers.get('X-User-Token')
+        user_id, app_uuid, data_source = parse_user_token(user_token)
+    except ValueError:
+        return api_error(410)
+
+    if request.method == 'PUT':
+        pin_json = request.json
+        if pin_json is None or pin_json.get('id') != pin_id:
+            return api_error(400)
+
+        pin = TimelinePin.query.filter_by(app_uuid=app_uuid, user_id=user_id, id=pin_id).one_or_none()
+        if pin is None:  # create pin
+            pin = TimelinePin.from_json(pin_json, app_uuid, user_id, data_source, 'web', [])
+            if pin is None:
+                return api_error(400)
+
+            user_timeline = UserTimeline(user_id=user_id,
+                                         type='timeline.pin.create',
+                                         pin=pin)
+            db.session.add(pin)
+            db.session.add(user_timeline)
+            db.session.commit()
+        else:  # update pin
+            try:
+                pin.update_from_json(pin_json)
+                user_timeline = UserTimeline(user_id=user_id,
+                                             type='timeline.pin.create',
+                                             pin=pin)
+                db.session.add(pin)
+                db.session.add(user_timeline)
+                db.session.commit()
+            except (KeyError, ValueError):
+                return api_error(400)
+
+    elif request.method == 'DELETE':
+        pin = TimelinePin.query.filter_by(app_uuid=app_uuid, user_id=user_id, id=pin_id).first_or_404()
+        user_timeline = UserTimeline(user_id=user_id,
+                                     type='timeline.pin.delete',
+                                     pin=pin)
+        db.session.add(user_timeline)
+        db.session.commit()
+    return 'OK'
+
+
+@api.route('/shared/pins/<pin_id>', methods=['PUT', 'DELETE'])
+def shared_pin(pin_id):
+    api_key = request.headers.get('X-API-Key')
+    return api_error(403)  # TODO: check api_key in dev portal, add/delete shared pins
```

Comment:
I think this is a tricky TODO!

Shared pins can have substantial fan-out (in the tens of thousands originally, maybe the hundreds for us), and when someone is subscribed to a topic they need to have all the pins (with related pin creation events) sent to them — though when unsubscribed, the mobile app is responsible for handling deletions, which makes that easier.

All that said, maybe we can punt on this: shared pins imply backend services that probably nobody is maintaining any more anyway. Leaving it as a TODO is fine for this iteration.